### ISSUE TITLE
Custom binding values are ignored in HTL files

### ISSUE THREAD

**Issue by henrykuijpers** [karollewandowski/aem-intellij-plugin#173]
We have a project where we have created a custom bindings value provider. Of course, AEM IntelliJ plugin now tells us that it does not know of this property. 

Is there a way to instruct the plugin that it exists?
I can think of a few things:

1. Class path scanning, to find BindingsValuesProvider classes and somehow find out what bindings they provide
2. Let users configure the plugin by adding their custom bindings + their types (because you probably need those to be able to provide completion)
3. ???

**Comment by karollewandowski** [comment]
I have this feature in private issue tracker and even started it, but switched to support for `data-sly-template`/`data-sly-call` for now and due to vacations period I have limited time for the project.
Approach I took at first is to register custom values via quick fix:
1. Plugin highlights variable as unreferenced variable.
2. User invokes fixes list and selects custom binding registration.
3. Popup shows up where user can provide binding type (optional, if not provided, there won't be completions for this binding).

Reason for that is that plugin can be used by WebStorm users which doesn't support Java and this solution will support both Java and non-Java IDEs. I plan to enhance it in future to automatically detect custom bindings, but other features have higher priority.

Custom binding variables will be implemented just after support for template/call statements, but I can't promise when it will be released. Support for template/call is almost done, but again - vacations...

**Comment by henrykuijpers** [comment]
Sounds awesome! 

I looked into automatically registering the binding value providers a little bit, but it's -- I believe -- too much of a hassle and there's no reliable way to find out what is being registered, without actually invoking the providers and somehow trying to see what objects they registered. 

There is no code contract that makes the providers tell the framework what objects they register (and even more importantly maybe, what type those objects are). 

I think your user interaction solution is a good way to approach it. Looking forward to see the change!

**Comment by karollewandowski** [comment]
At some point I'll implement connecting to AEM instance and there is a list of scripting variables exposed in: http://localhost:4502/system/console/scriptingvariables

I didn't test it yet for custom bindings, but I guess they will be listed there.

It would be also good to find variables in `BindingsValuesProvider` implementations for developers without local AEM instance. It would cover trivial cases like:
```java
bindings.put("myVar", variable);
```

If someone will add non-trivial variables like:
```java
bindings.put("var" + dynamicSuffix, variable);
```
then [s]he'll have to use manual quick fix to register variable. Maybe I will find some more edge cases to cover, we'll see.

**Comment by henrykuijpers** [comment]
/system/console/scriptingvariables is not working in my local 6.2 instance, is that correct?

**Comment by karollewandowski** [comment]
I'm using 6.3, maybe it's different in older versions:
<img width="279" alt="Screenshot 2019-08-11 at 20 19 01" src="https://user-images.githubusercontent.com/1154140/62837917-7d3f3280-bc75-11e9-9b97-de625ff12826.png">

**Comment by karollewandowski** [comment]
Feature implemented and will be available in 0.12 (which will be paid).
I decided to go with simplified approach - user can set custom HTL scripting variables in plugin settings and do it directly from editor via inspection fix:
![recording](https://i.imgur.com/NmTinqD.gif)

The reason is that custom variables are not used very often and so advanced mechanism is not a must for now. I will consider to implement it in future as improvement.
